### PR TITLE
[rush] Update rush NodeJS warning messages and add option to suppress non-LTS warning messages.

### PR DESCRIPTION
**Pull request by iclanton**
Fixes https://github.com/microsoft/web-build-tools/issues/1373

### CODE REVIEW

File: common/reviews/api/rush-lib.api.md [hunk 1]
```diff
@@ -221,10 +227,10 @@ export type ResolutionStrategy = 'fewer-dependencies' | 'fast';
 
 // @public
 export class Rush {
-    static launch(launcherVersion: string, isManaged: boolean): void;
-    static launchRushX(launcherVersion: string, isManaged: boolean): void;
+    static launch(launcherVersion: string, arg: ILaunchOptions): void;
+    static launchRushX(launcherVersion: string, arg: ILaunchOptions): void;
```
Comment (octogonz) [inline common/reviews/api/rush-lib.api.md]: The parameter should be called `options`.

File: apps/rush-lib/assets/rush-init/rush.json [hunk 2]
```diff
@@ -71,6 +71,19 @@
    */
   "nodeSupportedVersionRange": ">=10.13.0 <11.0.0",
```
Comment (octogonz) [inline apps/rush-lib/assets/rush-init/rush.json]: I tweaked the text for this option also.  I'm trying to replace `NodeJS` in our English text with `Node.js` because that's their official name.

File: apps/rush-lib/assets/rush-init/rush.json [hunk 3]
```diff
@@ -71,6 +71,19 @@
...
 
+  /**
+  * Odd-numbered major versions of Node.js are experimental.  Even-numbered releases
+  * spend six months in a stabilization period before the first Long Term Support (LTS) version.
+  * For example, 8.9.0 was the first LTS version of Node.js 8.  Pre-LTS versions are not recommended
+  * for production usage because they frequently have bugs.  They may cause Rush itself
+  * to malfunction.
+  *
+  * Rush normally prints a warning if it detects a non-LTS Node.js version.  If you are testing
```
Comment (octogonz) [inline apps/rush-lib/assets/rush-init/rush.json]: This should say `pre-LTS` instead of `non-LTS`, to avoid confusion with other warnings.

File: apps/rush-lib/src/api/Rush.ts [hunk 4]
```diff
@@ -63,12 +97,79 @@ export class Rush {
...
+  private static _warnAboutNodeVersion(rushConfiguration: RushConfiguration | undefined): void {
+    const nodeVersion: string = process.versions.node;
+
+    interface IExtendedNodeProcess extends NodeJS.Process {
+      release: {
+        lts?: string;
+      };
+    }
+
+    if (semver.satisfies(nodeVersion, '>= 11.0.0')) {
```
Comment (mikeharder) [inline apps/rush-lib/src/api/Rush.ts]: Should `suppressNodeLtsWarning` apply to this warning as well?  My preference would be for it to suppress both warnings.

Comment (octogonz) [reply apps/rush-lib/src/api/Rush.ts]: Let's discuss this in https://github.com/microsoft/web-build-tools/issues/1373 where the design notes are.

File: apps/rush-lib/src/api/Rush.ts [hunk 5]
```diff
@@ -63,12 +103,67 @@ export class Rush {
...
+      ));
+      console.log();
+    } else if (
+      rushConfiguration &&
+      !rushConfiguration.suppressNodeLtsWarning &&
+      !(process as IExtendedNodeProcess).release.lts
+    ) {
+      console.log();
+      console.warn(colors.yellow(
+        `Your version of Node.js (${nodeVersion}) has an odd major version number. ` +
```
Comment (octogonz) [inline apps/rush-lib/src/api/Rush.ts]: @bterlson pointed out that the meaning of "odd" is unclear here.

Let's change the wording to:  `Your version of Node.js (${nodeVersion}) is an odd-numbered release. `

**Comment by octogonz** [comment]
@iclanton Could you also test this using a nightly build (e.g. where `process.version` returns a string like `"v10.0.0-nightly20171114803cacd228"`)?  We should make sure an appropriate warning is printed in that case as well.

File: apps/rush-lib/src/api/Rush.ts [hunk 6]
```diff
@@ -3,12 +3,38 @@
...
 
 import { RushCommandLineParser } from '../cli/RushCommandLineParser';
 import { RushConstants } from '../logic/RushConstants';
 import { RushXCommandLine } from '../cli/RushXCommandLine';
 import { CommandLineMigrationAdvisor } from '../cli/CommandLineMigrationAdvisor';
+import { RushConfiguration } from './RushConfiguration';
+
+/**
+ * Options to pass to the rush "launch" functions.
+ */
```
Comment (octogonz) [inline apps/rush-lib/src/api/Rush.ts]: Needs `@public`

**Comment by mikeharder** [comment]
> @iclanton Could you also test this using a nightly build (e.g. where `process.version` returns a string like `"v10.0.0-nightly20171114803cacd228"`)? We should make sure an appropriate warning is printed in that case as well.

You could also use a nightly of `v13` if you want to test a version newer than `v12`.

File: apps/rush-lib/src/api/Rush.ts [hunk 7]
```diff
@@ -19,40 +45,54 @@ export class Rush {
...
    * This API is used by the `@microsoft/rush` front end to launch the "rush" command-line.
    * Third-party tools should not use this API.  Instead, they should execute the "rush" binary
-   * and start a new NodeJS process.
+   * and start a new Node.js process.
    *
    * @param launcherVersion - The version of the `@microsoft/rush` wrapper used to call invoke the CLI.
-   * @param isManaged - True if the tool was invoked from within a project with a rush.json file, otherwise false. We
-   *  consider a project without a rush.json to be "unmanaged" and we'll print that to the command line when
-   *  the tool is executed. This is mainly used for debugging purposes.
    */
```
Comment (octogonz) [inline apps/rush-lib/src/api/Rush.ts]: Add a `@remarks` note saying that earlier Rush front-ends used a different contract.  It's supported, but undocumented.

File: apps/rush-lib/src/api/Rush.ts [hunk 8]
```diff
@@ -63,12 +103,67 @@ export class Rush {
     return PackageJsonLookup.loadOwnPackageJson(__dirname).version;
   }
 
+  private static _tryGetRushConfiguration(): RushConfiguration | undefined {
+    try {
+      if (RushConfiguration.tryFindRushJsonLocation()) {
+        return RushConfiguration.loadFromDefaultLocation({ showVerbose: true });
+      }
+    } catch (e) {
+      // ignore
```
Comment (octogonz) [inline apps/rush-lib/src/api/Rush.ts]: This seems very similar to this other code:

https://github.com/microsoft/web-build-tools/blob/3fde77c427af0e4b58f0c7899734c899b4c8475c/apps/rush-lib/src/cli/RushCommandLineParser.ts#L113-L127

If I understand right, the underlying complication is that:

1. We prefer to print the version warnings as early as possible (to minimize any malfunctions), but...
2. `_warnAboutNodeVersion()` now needs to check `rushConfiguration.suppressNodeLtsWarning`, and...
3. Loading **rush.json** is complicated (and in the "unmanged" scenario it never gets loaded at all)

Your `_tryGetRushConfiguration()` is making a quick attempt to load **rush.json** for the purposes of `_warnAboutNodeVersion()`,  and then we come back back later and redo it with the more complete **rush.json** loading.

I'm not sure I like this approach.  It seems inefficient and potentially buggy.

Based on my conversation with @mikeharder, the purpose of the `suppressNodeLtsWarning` is merely to let people use `10.0.0` instead of `10.13.0`.  It does not affect the warnings about `6.15.0` or `9.11.2` or `98.0.0` or `10.0.0-nightly20171114803cacd228`.  Thus, the incompatibilities detected by the LTS warning are relatively benign.

I suggest to:

- Move (only) the `suppressNodeLtsWarning` warning to happen later, after the proper original **rush.json** configuration has been loaded.  This should untangle the `_tryGetRushConfiguration()` mess here.
- Let's consolidate all our versions as string constants in an `@internal` API (maybe **rush-lib/api/CompatibilityVersions.ts**?).  It can also be used by **rush** (via its non-version-selector **rush-lib** dependency).  This will make it easier for us to scatter the warning checks throughout the project in best places, while still having a nice central place to keep track of them all.

File: apps/rush/src/RushCommandSelector.ts [hunk 9]
```diff
@@ -7,6 +7,11 @@ import * as rushLib from '@microsoft/rush-lib';
 
 type CommandName = 'rush' | 'rushx' | undefined;
 
+export interface IExceuteOptions {
```
Comment (octogonz) [inline apps/rush/src/RushCommandSelector.ts]: Spelling.

Also, can't this be imported from the (non-version-selected) **rush-lib** dependency?

File: apps/rush-lib/src/api/Rush.ts (hunk 8 above)
Comment (mikeharder) [reply apps/rush-lib/src/api/Rush.ts]: @octogonz: I'm still not clear on the precise meaning of `suppressNodeLtsWarning`.  Could you or @iclanton create a small table showing which versions of Node should get which warning(s), and which of these warnings are suppressed by the setting?

My specific concern was that Node 12 Current (e.g. `12.6.0`) would get a warning by default, and `suppressNodeLtsWarning` would be needed to suppress it.  I want to ensure our repo can use all LTS and even-numbered "Current" versions of Node with no warnings.

Comment (octogonz) [reply apps/rush-lib/src/api/Rush.ts]: Sorry about the confusion.  I used `10.0.0` as my example because it has a corresponding LTS version `10.13.0`.  The `12.x` version series doesn't have an LTS version number yet, so I was trying to avoid using it.

Let me rephrase like this:

The purpose of the `suppressNodeLtsWarning` is merely to let people use e.g. `<even>.x.x` instead of `<even>.<LTS>.x`. 

It does not affect the warnings about:
- `6.15.0` (an outdated release), or
- `9.11.2` (an `<odd>.x.x` release), or
- `98.0.0` (a "too new" release), or
- `10.0.0-nightly20171114803cacd228` (a nightly release)

File: apps/rush-lib/src/api/Rush.ts (hunk 4 above)
Comment (iclanton) [reply apps/rush-lib/src/api/Rush.ts]: Discussed in the issue. Consensus is that we don't suppress the warning for any odd-numbered releases.

Comment (octogonz) [reply apps/rush-lib/src/api/Rush.ts]: Right, but I think Mike's point was that `>= 11.0.0` will match `12.0.0` as well, which is not intended.

This number should probably be increased to `>= 13.0.0`.

File: apps/rush-lib/src/api/Rush.ts [hunk 10]
```diff
@@ -63,12 +103,67 @@ export class Rush {
...
+      console.log();
+      console.warn(colors.yellow(
+        `Your version of Node.js (${nodeVersion}) has not been tested with this release` +
+        `of the Rush engine.  Please consider upgrading the "rushVersion" setting in rush.json, ` +
+        `or  downgrading Node.js.`
+      ));
+      console.log();
+    } else if (
+      rushConfiguration &&
+      !rushConfiguration.suppressNodeLtsWarning &&
```
Comment (octogonz) [inline apps/rush-lib/src/api/Rush.ts]: The `suppressNodeLtsWarning` option should not apply to the "odd major version" warning.  We always want to show the "odd major version" warning.

File: apps/rush/src/start.ts [hunk 11]
```diff
@@ -1,38 +1,32 @@
...
-    + ` The Rush team will not accept issue reports for it.`
-    + ` Please consider upgrading Rush or downgrading Node.js.`));
-}
-
-// We are not on an LTS release
-// tslint:disable-next-line
-else if (!semver.satisfies(nodeVersion, '^6.9.0')
-      && !semver.satisfies(nodeVersion, '^8.9.0')
-      && !semver.satisfies(nodeVersion, '^10.13.0')) {
-  console.warn(colors.yellow(`Your version of Node.js (${nodeVersion}) is not a Long-Term Support (LTS) release.`
```
Comment (octogonz) [inline apps/rush/src/start.ts]: Seems like the "not LTS" message was removed entirely.  We need to re-add it in **rush-lib**.  Per my other comment, it should probably be performed /after/ rush.json has been loaded, so we can consider the `suppressNodeLtsWarning` option.

File: apps/rush-lib/src/api/Rush.ts (hunk 4 above)
Comment (iclanton) [reply apps/rush-lib/src/api/Rush.ts]: Yeah I made that change as well. Take a look at the latest changes.

**Comment by iclanton** [comment]
This isn't quite ready yet. Still working on it.

File: apps/rush-lib/src/api/RushConfiguration.ts [hunk 12]
```diff
@@ -775,6 +777,23 @@ export class RushConfiguration {
...
+    * spend six months in a stabilization period before the first Long Term Support (LTS) version.
+    * For example, 8.9.0 was the first LTS version of Node.js 8.  Pre-LTS versions are not recommended
+    * for production usage because they frequently have bugs.  They may cause Rush itself
+    * to malfunction.
+    *
+    * Rush normally prints a warning if it detects a pre-LTS Node.js version.  If you are testing
+    * pre-LTS versions in preparation for supporting the first LTS version, you can use this setting
+    * to disable Rush's warning.
+    *
+    * @beta
```
Comment (octogonz) [inline apps/rush-lib/src/api/RushConfiguration.ts]: I'm not sure this needs to be marked as `@beta`.  The rush.json schema change will impact compatibility.

File: apps/rush-lib/src/cli/RushCommandLineParser.ts [hunk 13]
```diff
@@ -35,23 +35,29 @@ import { GlobalScriptAction } from './scriptActions/GlobalScriptAction';
...
 
 /**
  * Options for `RushCommandLineParser`.
  */
 export interface IRushCommandLineParserOptions {
-  cwd?: string;   // Defaults to `cwd`
+  cwd: string;   // Defaults to `cwd`
+
+  /**
+   * @internal
```
Comment (octogonz) [inline apps/rush-lib/src/cli/RushCommandLineParser.ts]: `@internal` has no meaning here, since `IRushCommandLineParserOptions` is private to this NPM package.

File: apps/rush-lib/src/cli/RushCommandLineParser.ts [hunk 14]
```diff
@@ -35,23 +35,29 @@ import { GlobalScriptAction } from './scriptActions/GlobalScriptAction';
...
   public telemetry: Telemetry | undefined;
-  public rushConfiguration: RushConfiguration;
   public rushGlobalFolder: RushGlobalFolder;
+  public rushConfiguration: RushConfiguration;
 
   private _debugParameter: CommandLineFlagParameter;
   private _rushOptions: IRushCommandLineParserOptions;
 
-  constructor(options?: IRushCommandLineParserOptions) {
+  constructor(options?: Partial<IRushCommandLineParserOptions>) {
```
Comment (octogonz) [inline apps/rush-lib/src/cli/RushCommandLineParser.ts]: As far as I know, there's only one line of code that constructs RushCommandLineParser().  Why are the constructor parameters ultra-optional?  Shouldn't every parameter simply be required?

File: apps/rush-lib/src/cli/RushCommandLineParser.ts [hunk 15]
```diff
@@ -63,11 +69,27 @@ export class RushCommandLineParser extends CommandLineParser {
...
+      if (rushJsonFilename) {
+        this.rushConfiguration = RushConfiguration.loadFromConfigurationFile(rushJsonFilename);
+      }
+    } catch (error) {
+      this._reportErrorAndSetExitCode(error);
+    }
+
     this._populateActions();
+
+    NodeJsCompatibility.warnAboutCompatibilityIssues(
```
Comment (octogonz) [inline apps/rush-lib/src/cli/RushCommandLineParser.ts]: Should this precede `_populateActions()`?

File: apps/rush-lib/src/cli/RushCommandLineParser.ts (hunk 14 above)
Comment (octogonz) [reply apps/rush-lib/src/cli/RushCommandLineParser.ts]: That would also eliminate the duplicated `_normalizeOptions()`

Comment (iclanton) [reply apps/rush-lib/src/cli/RushCommandLineParser.ts]: I'd rather keep this contract mostly the same in this PR. The `Partial` keeps the constructor contract the same while making the internals simpler.

File: apps/rush-lib/src/utilities/Utilities.ts [hunk 16]
```diff
@@ -722,7 +722,7 @@ export class Utilities {
 
     if (result.error && (result.error as any).errno === 'ENOENT') { // tslint:disable-line:no-any
       // This is a workaround for GitHub issue #25330
-      // https://github.com/nodejs/node-v0.x-archive/issues/25330
+      // https://github.com/Node.js/node-v0.x-archive/issues/25330
```
Comment (octogonz) [inline apps/rush-lib/src/utilities/Utilities.ts]: Bad search+replace

File: apps/rush-lib/src/cli/RushCommandLineParser.ts (hunk 15 above)
Comment (iclanton) [reply apps/rush-lib/src/cli/RushCommandLineParser.ts]: In most cases it shouldn't matter, but I'll move it for the edge cases.

File: apps/rush/src/start.ts [hunk 17]
```diff
@@ -1,39 +1,19 @@
...
-    + ` These versions frequently contain bugs, and the Rush team will not accept issue reports for them.`
-    + ` Please consider installing a stable release.`));
-}
+const alreadyReportedNodeTooNewError: boolean = NodeJsCompatibility.warnAboutVersionTooNew(false, false);
```
Comment (octogonz) [inline apps/rush/src/start.ts]: `warnAboutVersionTooNew(false, false)` is not a very readable API.  What is false?  What else isn't false?

You could probably use `ILaunchOptions` to describe this, especially if you get rid of that persnickety optional member.

File: apps/rush/src/start.ts [hunk 18]
```diff
@@ -1,39 +1,19 @@
...
 // See LICENSE in the project root for license information.
 
-// The minimal set of imports that are safe even for ancient NodeJS versions:
+// The minimal set of imports that are safe even for ancient Node.js versions:
 import * as colors from 'colors';
 import * as os from 'os';
 import * as semver from 'semver';
 
-const nodeVersion: string = process.versions.node;
+import { NodeJsCompatibility } from '@microsoft/rush-lib/lib/logic/NodeJsCompatibility';
```
Comment (octogonz) [inline apps/rush/src/start.ts]: This should really be marked as `@internal` rather than accessed via a path-based import.  That's what `@internal` is for.  We should ideally work to eliminate all the path-based imports from our code base.

Comment (iclanton) [reply apps/rush/src/start.ts]: I used a path-based import in this one case because this code runs before we eval the other (more complex) imports.

The reasoning is that we should provide a meaningful error message before the tool crashes if someone is on a really old NodeJS version that, for whatever reason, fails to evaluate the imports.

Comment (octogonz) [reply apps/rush/src/start.ts]: That's persuasive.  Maybe add a comment explaining it.

File: apps/rush/src/start.ts (hunk 17 above)
Comment (iclanton) [reply apps/rush/src/start.ts]: The `isManaged` property doesn't really correctly apply here. I'll put together an options object, though.
